### PR TITLE
[WB-9234] Raise when Table size exceeds limit

### PR DESCRIPTION
**Pull request by dannygoldstein**
Fixes https://wandb.atlassian.net/browse/WB-9234

Description
-----------
Currently if a user logs a dataset that is more than 200k rows long as a `wandb.Table`, we (after warning, but not raising an exception) only log the first 200k rows to W&B. This can lead to unexpected behavior in the frontend where after a run is over a user finds that only part of their dataset is logged. This PR updates the behavior to throw an exception in this case, and gives users a pointer for how they can override the default 200k row limit. 

Testing
-------
Unit test

### CODE REVIEW

File: wandb/data_types.py
Comment (hu-po) [inline wandb/data_types.py:508]: Maybe remove this documentation line since no `warn` kwarg exists anymore?

Comment (dannygoldstein) [reply wandb/data_types.py]: nice - thanks. done

**Review by tssweeney — APPROVED** [review]
This looks correct, but I am a bit concerned about actually doing this. Are we sure we want to crash the user script in these cases? I feel like we have heard many times from customers that we should be soft warning on things like this. What if this is in a big training script and they lose their work? Can we get a firm approval from @cvphelps ? Carey: the implication here is that if a user logs a table > 200k rows - their job will crash.

**Review by tssweeney — CHANGES_REQUESTED** [review]
blocking until we get hard approval from Carey

**Comment by jlzhao27** [comment]
@tssweeney do we have any access to sdk settings/env vars in that block? we could condition the failures on an env var like `TABLES_EXIT_ON_FAILURE` or something

File: wandb/data_types.py
Comment (dannygoldstein) [inline wandb/data_types.py:239]: @dmitryduev should we be doing this with settings? is there a way to?

Comment (dmitryduev) [reply wandb/data_types.py:239]: great idea to do it that way, I'll explain.

**Comment by raubitsj** [comment]
> This looks correct, but I am a bit concerned about actually doing this. Are we sure we want to crash the user script in these cases? I feel like we have heard many times from customers that we should be soft warning on things like this. What if this is in a big training script and they lose their work? Can we get a firm approval from @cvphelps ? Carey: the implication here is that if a user logs a table > 200k rows - their job will crash.

We have been leaning towards not crashing in these instances, instead we should warn (ideally inline, in the footer, and in the UI) and we have a strict mode (not publicized and possibly not even used yet) which can turn warnings into errors (WANDB_STRICT=true and associated settings)

**Comment by jlzhao27** [comment]
Thanks for the comment Jeff, we had the same discussion here: https://weightsandbiases.slack.com/archives/C032F05V5HS/p1649869538076629

I think for this particular case, the warning is a good start and this user is asking specifically for a crash/strict mode. I agree with not crashing long running training jobs but tables are also during dataset ETL jobs and logging partial data should be a catastrophic failure since it affects the integrity of the dataset.

I'm in favor of this change as long as we rely on best practices to toggle the fail hard mode.

**Comment by dannygoldstein** [comment]
Should we just tell the user to try `WANDB_STRICT=true`?

**Comment by jlzhao27** [comment]
I like the general approach of `WANDB_STRICT=true`, im just worried that there are warning messages that are spammy. Do you have any sense of how common `warning` messages show up? Can we introduce a stronger `error` state and a setting to make that fail hard?

**Comment by dannygoldstein** [comment]
@jlzhao27 @raubitsj @dmitryduev @tssweeney @cvphelps 

I re-implemented this using settings (thanks for help @dmitryduev), now there is a sdk setting you can set to opt-in to. If you opt-in, it will raise an exception if the length of the table exceeds `MAX_ARTIFACT_ROWS`. We should also update the documentation to tell people to modify `MAX_ARTIFACT_ROWS` instead of `MAX_ROWS` which is currently there 

So we can tell JnJ to just do 

`WANDB_TABLE_RAISE_ON_MAX_ROW_LIMIT_EXCEEDED=True`

and they will get the behavior they want

**Review by jlzhao27 — APPROVED** [review]
lgtm from my end, should wait for a SDK review

File: tests/test_data_types.py
Comment (dmitryduev) [inline tests/test_data_types.py:594]: Could you please replace this and below with 
```
with mock.patch("wandb.Table.MAX_ARTIFACT_ROWS", 10):
```
We prefer avoiding relying on `finally` statements. Also, this will reset this attr globally for the pytest worker, so there may be unexpected future test failures (e.g. if a new test is added where this is assumed to have a different value).

**Review by dmitryduev — COMMENTED** [review]
Thanks @dannygoldstein. In general this LGTM, but I'd like to hear your thoughts on the following before giving it my final thumbs-up.
This issue seems super specific to a particular scenario, and we're not sure that the way to control the SDK behavior in this case (and similar) is by introducing a super-specific knob in the settings. Instead, how about displaying a warning by default and raising an error only if `wandb.run.settings.strict is True`? I don't think we use it much in the codebase rn (if at all?..), but that was exactly the intent of having such a setting. [So the user would enable the strict mode in case they want us to crash their script in such a case.]
As a side note, we have many more users complaining (and righteously so!) about us actually crashing their runs at times instead of switching to a noop mode, so we're striving to eliminate such scenarios as the default behavior and instead move hard failures under the `strict` flag.

File: tests/test_data_types.py
Comment (dannygoldstein) [reply tests/test_data_types.py]: done

**Comment by dannygoldstein** [comment]
> Thanks @dannygoldstein. In general this LGTM, but I'd like to hear your thoughts on the following before giving it my final thumbs-up. 

> This issue seems super specific to a particular scenario, and we're not sure that the way to control the SDK behavior in this case (and similar) is by introducing a super-specific knob in the settings. Instead, how about displaying a warning by default and raising an error only if `wandb.run.settings.strict is True`? 

@dmitryduev See @jlzhao27's comment above. Right now the default behavior is to warn, and we preserve that here. But we worry that there may be other warnings that may trigger crashing behavior to be too aggressive if we enable `WANDB_STRICT=true`. This way we can ensure that we will only crash on this one specific warning, which is what the customer wants.

**Review by tssweeney — APPROVED** [review]

**Review by dmitryduev — APPROVED** [review]
👍🏻

**Pull request merged**
